plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveType.java

```diff
@@ -236,15 +257,31 @@ public List<String> getHiveDereferenceNames(List<Integer> dereferences)
         ImmutableList.Builder<String> dereferenceNames = ImmutableList.builder();
         TypeInfo typeInfo = getTypeInfo();
         for (int fieldIndex : dereferences) {
-            checkArgument(typeInfo instanceof StructTypeInfo, "typeInfo should be struct type", typeInfo);
-            StructTypeInfo structTypeInfo = (StructTypeInfo) typeInfo;
-
             checkArgument(fieldIndex >= 0, "fieldIndex cannot be negative");
-            checkArgument(fieldIndex < structTypeInfo.getAllStructFieldNames().size(),
-                    "fieldIndex should be less than the number of fields in the struct");
-            String fieldName = structTypeInfo.getAllStructFieldNames().get(fieldIndex);
-            dereferenceNames.add(fieldName);
-            typeInfo = structTypeInfo.getAllStructFieldTypeInfos().get(fieldIndex);
+
+            if (typeInfo instanceof StructTypeInfo structTypeInfo) {
+                checkArgument(fieldIndex < structTypeInfo.getAllStructFieldNames().size(),
+                        "fieldIndex should be less than the number of fields in the struct");
+
+                String fieldName = structTypeInfo.getAllStructFieldNames().get(fieldIndex);
+                dereferenceNames.add(fieldName);
+                typeInfo = structTypeInfo.getAllStructFieldTypeInfos().get(fieldIndex);
+            }
+            else if (typeInfo instanceof UnionTypeInfo unionTypeInfo) {
+                checkArgument((fieldIndex - 1) < unionTypeInfo.getAllUnionObjectTypeInfos().size(),
+                        "fieldIndex should be less than the number of fields in the union plus tag field");
+
+                if (fieldIndex == 0) {
+                    dereferenceNames.add("tag");
+                }
+                else {
+                    typeInfo = unionTypeInfo.getAllUnionObjectTypeInfos().get(fieldIndex - 1);
+                    dereferenceNames.add("field" + (fieldIndex - 1));
```

Comment:
I'd add a constant/methods in HiveTypeTranslator to make sure the strings "tag" and "fieldX" come from one place and don't diverge

Reply:
ok